### PR TITLE
Artist storage

### PR DESCRIPTION
**Pull request by artemstorozhuk**

### CODE REVIEW

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.m14n.billibmemory
+
+class ArtistStorage(
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:3]: Add interface with docs

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    artists: Collection<Artist>
+) {
+
+    private val artists: Collection<Artist> = ArrayList(artists)
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:7]: Why can't you use original data? Isn't copy a performance and memory overhead?

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    private val idToArtist: Map<Long, Artist> = HashMap(
+        artists.asSequence()
+            .map { it.id to it }
+            .toMap())
+    private val nameToArtist: Map<String, Artist> = HashMap(
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:12]: can be replaced with `artists.associateBy(it.name, it)`

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        artists.asSequence()
+            .map { it.name to it }
+            .toMap())
+
+    fun getAll() = artists
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:17]: May be to return a copy of that collection more reasonable.

File: src/test/kotlin/com/m14n/billibmemory/ArtistStorageTest.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    private val artists = listOf(
+        Artist(1, "Katy Perry"),
+        Artist(2, "Madonna"),
+        Artist(3, "Taylor Swift")
+    )
+
+    @Test
+    fun `get all should return all arists`() {
+        ArtistStorage(artists)
```
Comment (adjorno) [inline src/test/kotlin/com/m14n/billibmemory/ArtistStorageTest.kt:17]: move SUT creation to `@Before` preparation function

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt (hunk 2 above)
Comment (artemstorozhuk) [reply src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:7]: Because the client may modify input list. This copy prevents such behaviour.

Comment (artemstorozhuk) [reply src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:7]: I will add such test to make it more clear.

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt (hunk 4 above)
Comment (artemstorozhuk) [reply src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:17]: This method returns an unmodifiable collection, so the client can't modify it unless casting to the specific list is performed.

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt (hunk 2 above)
Comment (adjorno) [reply src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:9]: Still the overhead is here and for huge amount of data may be substantial.
I'd leave it up to the user to decide. If user wants to work to the copy, just pass data.copy() to the constructor.

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt (hunk 4 above)
Comment (adjorno) [reply src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt]: This is not clear from the implementation. Just wrap it into unmodifiable collection explicitly by calling Collection::toReadOnly()

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt (hunk 2 above)
Comment (adjorno) [reply src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:9]: Just make it clear in docs. It's not good to manipulate with input data.

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    FindByIdStorage<Artist>,
+    FindByNameStorage<Artist> {
+
+    private val artists: Collection<Artist> = ArrayList(artists)
+    private val idToArtist: Map<Long, Artist> =
+        artists.associateBy { it.id }
+    private val nameToArtist: Map<String, Artist> =
+        artists.associateBy { it.name }
+
+    override fun getAll() = artists
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:15]: can be delegated to generic implementation

File: src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+    override fun findById(id: Long) = idToArtist[id]
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/ArtistStorage.kt:17]: Can be delegated to generic implementation

File: src/main/kotlin/com/m14n/billibmemory/CollectionStorage.kt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package com.m14n.billibmemory
+
+interface CollectionStorage<T> {
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/CollectionStorage.kt:3]: Can have generic implementation

File: src/main/kotlin/com/m14n/billibmemory/FindByIdStorage.kt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package com.m14n.billibmemory
+
+interface FindByIdStorage<T> {
```
Comment (adjorno) [inline src/main/kotlin/com/m14n/billibmemory/FindByIdStorage.kt:3]: Most of entities have ids. Makes sense to come up with generic interface HasId and implementation

**Pull request closed without merging**